elastic/kibana
# [Dashboard] Read App State from URL on Soft Refresh

## Summary

This PR makes dashboard subscribe to URL changes and overwrite its state with any state added in via the URL. This way it is capable of updating state directly from the URL without a hard refresh.

Fixes https://github.com/elastic/kibana/issues/93685
Fixed https://github.com/elastic/kibana/issues/108548

### How to test
**Test State Load With Soft Refresh**
1. Create a new dashboard
2. Add a filter, customize the time range
3. Write anything into the query bar
4. Paste in the following URL (editing it to your environment) `{yourKibanaURL}/app/dashboards#/create?_a=(query:(language:kuery,query:'testing hello'))`. 
5. The browser should not do a hard refresh
6. "testing hello" should replace the value you wrote in step 3
7. The filters and time range you configured in step 2 should still be present

**Test back button loads configured dashboard time range**
1. Load a dashboard which has time restore (the ecommerce revenue dashboard for instance).
2. Notice the default time range
3. Change the time range to something else
4. Press the browser back button. The dashboard should go back to the time range from step 2 rather than the default value (last 15 minutes).

**Snapshot sharing**
1. Create a snapshot share from any dashboard
2. Change the state in as many ways as you want
3. paste the snapshot URL back into the browser
4. The dashboard should reset back to the snapshot state without a hard refresh.


### Checklist

Delete any items that are not applicable to this PR.
